### PR TITLE
Fixes for viewport selection issues

### PR DESCRIPTION
**Pull request by hultonha**
Update:

`ViewportScreenToWorld/WorldToScreen` code has been factored out of `RenderViewportWidget` and tests have been added for it (trying to instantiate `RenderViewportWidget` in a fixture turned out to be a total nightmare 😞).

This change cleans up quite a few bits and bobs (old code and unifying existing code). It also puts us in good shape to update the internal implementations should we choose to as we have tests covering the expected (correct) behavior.

```bash
[==========] Running 5 tests from 2 test cases.
[----------] Global test environment set-up.
[----------] 1 test from AtomToolsFrameworkTest
[ RUN      ] AtomToolsFrameworkTest.GetExteralReferencePath_Succeeds
[       OK ] AtomToolsFrameworkTest.GetExteralReferencePath_Succeeds (6 ms)
[----------] 1 test from AtomToolsFrameworkTest (6 ms total)

[----------] 4 tests from ViewportInteractionImplFixture
[ RUN      ] ViewportInteractionImplFixture.ViewportInteractionRequestsMapsFromScreenToWorldAndBack
[       OK ] ViewportInteractionImplFixture.ViewportInteractionRequestsMapsFromScreenToWorldAndBack (1 ms)
[ RUN      ] ViewportInteractionImplFixture.ScreenToWorldReturnsPositionOnNearClipPlaneInWorldSpace
[       OK ] ViewportInteractionImplFixture.ScreenToWorldReturnsPositionOnNearClipPlaneInWorldSpace (0 ms)
[ RUN      ] ViewportInteractionImplFixture.WorldToScreenGivesExpectedScreenCoordinates
[       OK ] ViewportInteractionImplFixture.WorldToScreenGivesExpectedScreenCoordinates (0 ms)
[ RUN      ] ViewportInteractionImplFixture.ScreenToWorldRayGivesGivesExpectedOriginAndDirection
[       OK ] ViewportInteractionImplFixture.ScreenToWorldRayGivesGivesExpectedOriginAndDirection (0 ms)
[----------] 4 tests from ViewportInteractionImplFixture (2 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 2 test cases ran. (8 ms total)
[  PASSED  ] 5 tests.
```

This PR fixes a couple of small but irritating issues with selection in the viewport.

The first relates to the origin of the ray for mouse clicks not actually starting on the near clip plane. This can cause awkward situations where it appears as though the user should be able to click something, but a quad is obscuring the collision in between the camera position and near clip plane position.

https://user-images.githubusercontent.com/82228511/141158369-29371732-f169-42c6-91c2-1398d4617db8.mp4

The other issue that is fixed relates to icons still being clickable/intersectable even when not visible (as they are part of closed entity containers).

I've also removed the old `PickEntity` from `EditorViewportWidget` which is no longer used and relies on legacy calls.

## Update

Initial test added:

before:
```
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from EditorTransformComponentSelectionViewportPickingManipulatorTestFixture
[ RUN      ] EditorTransformComponentSelectionViewportPickingManipulatorTestFixture.BoundsBetweenCameraAndNearClipPlaneDoesNotIntersectMouseRay
D:/o3de-2/Code/Framework/AzToolsFramework/Tests/EditorTransformComponentSelectionTests.cpp(971): error: Value of: selectedEntities
Expected: has 1 element and that element is equal to 8-byte object <20-18 26-03 0E-00 00-00>
  Actual: { 8-byte object <20-18 26-03 0C-00 00-00> }, where the following matchers don't match any elements:
matcher #0: is equal to 8-byte object <20-18 26-03 0E-00 00-00>
and where the following elements don't match any matchers:
element #0: 8-byte object <20-18 26-03 0C-00 00-00>
[  FAILED  ] EditorTransformComponentSelectionViewportPickingManipulatorTestFixture.BoundsBetweenCameraAndNearClipPlaneDoesNotIntersectMouseRay (257 ms)
[----------] 1 test from EditorTransformComponentSelectionViewportPickingManipulatorTestFixture (259 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (278 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] EditorTransformComponentSelectionViewportPickingManipulatorTestFixture.BoundsBetweenCameraAndNearClipPlaneDoesNotIntersectMouseRay
```

after:
```
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from EditorTransformComponentSelectionViewportPickingManipulatorTestFixture
[ RUN      ] EditorTransformComponentSelectionViewportPickingManipulatorTestFixture.BoundsBetweenCameraAndNearClipPlaneDoesNotIntersectMouseRay
[       OK ] EditorTransformComponentSelectionViewportPickingManipulatorTestFixture.BoundsBetweenCameraAndNearClipPlaneDoesNotIntersectMouseRay (303 ms)
[----------] 1 test from EditorTransformComponentSelectionViewportPickingManipulatorTestFixture (305 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (328 ms total)
[  PASSED  ] 1 test.
```

Note: This doesn't actually test the code paths in EditorLib but it is a useful example using the manipulator test framework. Need to add more coverage in `EditorLib.Tests`.

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportSelection/EditorVisibleEntityDataCache.h [hunk 1]
```diff
@@ -55,7 +55,11 @@ namespace AzToolsFramework
         bool IsVisibleEntityVisible(size_t index) const;
         bool IsVisibleEntitySelected(size_t index) const;
         bool IsVisibleEntityIconHidden(size_t index) const;
-        bool IsVisibleEntitySelectableInViewport(size_t index) const;
+        //! Returns true if the entity is individually selectable (not inside an entity container)
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ViewportSelection/EditorVisibleEntityDataCache.h:58]: nit: "not inside an entity container" doesn't seem like quite the right description. I think "its nearest container entity ancestor is being focused" is more accurate but maybe @AMZN-daimini could weigh in.

**Review by NicholasVanSickle — APPROVED** [review]
Change looks good to me so approving, have a bit of non-blocking feedback

File: Code/Editor/ViewportManipulatorController.cpp [hunk 2]
```diff
@@ -117,16 +119,12 @@ namespace SandboxEditor
                     aznumeric_cast<int>(position->m_normalizedPosition.GetX() * windowSize.m_width),
                     aznumeric_cast<int>(position->m_normalizedPosition.GetY() * windowSize.m_height));
 
-                m_mouseInteraction.m_mousePick.m_screenCoordinates = screenPoint;
-                AZStd::optional<ProjectedViewportRay> ray;
-                ViewportInteractionRequestBus::EventResult(
-                    ray, GetViewportId(), &ViewportInteractionRequestBus::Events::ViewportScreenToWorldRay, screenPoint);
```
[the remark this answers is not in the document]
Comment (hultonha) [reply Code/Editor/ViewportManipulatorController.cpp:123]: Thanks for the feedback Nick! I like the idea of cleaning things up so I'll see if I can make the changes you suggest. We might be able to eliminate some of these EBus calls entirely and push people towards using the function `AzFramework::ScreenToWorld`/`WorldToScreen` with the relevant `CameraState` they can request (and then cache to call the overloads of `ScreenToWorld`/`WorldToScreen` that take relevant matrices. Thanks for looking into the callsites, I'll update the PR as soon as possible first thing tomorrow. Cheers!

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportSelection/EditorVisibleEntityDataCache.h (hunk 1 above)
Comment (hultonha) [reply Code/Framework/AzToolsFramework/AzToolsFramework/ViewportSelection/EditorVisibleEntityDataCache.h:58]: If you pass this an `EntityId` of a 'contained' entity then it won't be technically individually selectable (hence the name change), but I could try and phrase this in an easier to understand way. If you have any suggestions @AMZN-daimini please let me know. Thanks!

Comment (AMZN-daimini) [reply Code/Framework/AzToolsFramework/AzToolsFramework/ViewportSelection/EditorVisibleEntityDataCache.h:58]: Container Entity is an Editor concept now, and that's what we're referring to here. (The Prefab Container, incidentally, is also a Container Entity, but this code isn't explicitly aware of Prefabs, and the Container Entity concept can be used in other circumstances in the future).

To make this comment clearer, I would reword it like this:
```
//! Returns true if the entity is individually selectable (none of its ancestors is a closed container entity).
//! @note It may still be desirable to be able to 'click' an entity that is a descendant of a closed container
//! to select the container itself, not the individual entity.
```

Comment (hultonha) [reply Code/Framework/AzToolsFramework/AzToolsFramework/ViewportSelection/EditorVisibleEntityDataCache.h]: Comment is now updated

File: Code/Editor/ViewportManipulatorController.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (hultonha) [reply Code/Editor/ViewportManipulatorController.cpp:123]: @NicholasVanSickle - I've made some more changes to the viewport picking functions to simplify the API and leverage the existing picking code.

I have removed the depth parameter from `ScreenToWorld` as I believe it's actually not particularly helpful. As depth isn't linear it's incredibly difficult to know what a value in the `0-1` range will actually map to. I see the only place it was actually used was in some legacy code to get the near/far plane position, but the ray can be calculated more easily just be getting the near plane position and subtracting it from the camera position. I've updated the old call to reflect this (which should ultimately be removed anyway, nothing I know of is actually using that right now).

I'm going to look at adding some integration tests to validate the API (do any exist already?). I haven't found any issues through adhoc testing.

Please let me know if you have any thoughts/concerns. I'll be looking at tests next

**Review by jonawals — APPROVED** [review]
Only looked at the AzManipulatorTestFramework code but LGTM 👍

**Review by cgalvan — APPROVED** [review]

**Review by AMZN-Igarri — APPROVED** [review]

**Review by cgalvan — APPROVED** [review]

File: Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    {
+        using AzFramework::ScreenPoint;
+
+        m_view->SetCameraTransform(AZ::Matrix3x4::CreateFromMatrix3x3AndTranslation(
+            AZ::Matrix3x3::CreateRotationZ(AZ::DegToRad(90.0f)), AZ::Vector3(10.0f, 0.0f, 5.0f)));
+
+        {
+            const auto expectedScreenPoint = ScreenPoint{ 600, 450 };
+            const auto resultScreenPoint = ScreenToWorldToScreen(expectedScreenPoint, *m_viewportInteractionImpl);
+            EXPECT_EQ(resultScreenPoint, expectedScreenPoint);
```
Comment (burelc-amzn) [inline Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp:92]: The output of this when the test fails is this:
```
/home/ANT.AMAZON.COM/burelc/ws/scratch/Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp:92: Failure
Expected equality of these values:
  resultScreenPoint
    Which is: 8-byte object <2F-02 00-00 C2-01 00-00>
  expectedScreenPoint
    Which is: 8-byte object <58-02 00-00 C2-01 00-00>
```

We can improve it a bit by adding a function to make a matcher for that `ScreenPoint` type:
```cpp
auto MatchesScreenPoint(const AzFramework::ScreenPoint& screenPoint)
{
    return testing::AllOf(
        testing::Field(&AzFramework::ScreenPoint::m_x, testing::Eq(screenPoint.m_x)),
        testing::Field(&AzFramework::ScreenPoint::m_y, testing::Eq(screenPoint.m_y))
    );
}

EXPECT_THAT(resultScreenPoint, MatchesScreenPoint(expectedScreenPoint));
```

Then the output is:
```
/home/ANT.AMAZON.COM/burelc/ws/scratch/Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp:100: Failure
Value of: resultScreenPoint
Expected: (is an object whose given field is equal to 600) and (is an object whose given field is equal to 450)
  Actual: 8-byte object <2F-02 00-00 C2-01 00-00> (of type AzFramework::ScreenPoint), whose given field is 559 (of type int)
```

We can make that better again by providing a `PrintTo` overload for `ScreenPoint`:
```cpp
namespace AzFramework
{
    void PrintTo(const ScreenPoint& point, std::ostream* os)
    {
        *os << "{x: " << point.m_x << ", y: " << point.m_y << "}";
    }
} // namespace AzFramework
```

Now the output is:
```
/home/ANT.AMAZON.COM/burelc/ws/scratch/Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp:108: Failure
Value of: resultScreenPoint
Expected: (is an object whose given field is equal to 600) and (is an object whose given field is equal to 450)
  Actual: {x: 559, y: 450} (of type AzFramework::ScreenPoint), whose given field is 559 (of type int)
```

I'd advocate for that `PrintTo` function to be in some header inside `AzFramework` itself, similar to how we have that `IsClose` matcher template inside `Code/Framework/AzCore/Tests/AZTestShared/Math/MathTestHelpers.h`

File: Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        }
+
+        {
+            auto expectedScreenPoint = ScreenCenter();
+            const auto resultScreenPoint = ScreenToWorldToScreen(expectedScreenPoint, *m_viewportInteractionImpl);
+            EXPECT_EQ(resultScreenPoint, expectedScreenPoint);
+        }
```
Comment (burelc-amzn) [inline Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp:99]: Nit, repeated: I always prefer to have tests call the function under test only once. If you have multiple data sets that you want to test, use a parameterized fixture.

File: Gems/Atom/RPI/Code/Source/RPI.Public/View.cpp [hunk 5]
```diff
@@ -47,17 +47,20 @@ namespace AZ
...
+            // Set default matrices 
             SetWorldToViewMatrix(AZ::Matrix4x4::CreateIdentity());
             AZ::Matrix4x4 viewToClipMatrix;
             AZ::MakePerspectiveFovMatrixRH(viewToClipMatrix, AZ::Constants::HalfPi, 1, 0.1f, 1000.f, true);
             SetViewToClipMatrix(viewToClipMatrix);
 
-            Data::Asset<ShaderAsset> viewSrgShaderAsset = RPISystemInterface::Get()->GetCommonShaderAssetForSrgs();
-
-            if (viewSrgShaderAsset.IsReady())
+            if (auto rpiSystemInterface = RPISystemInterface::Get())
```
Comment (VickyAtAZ) [inline Gems/Atom/RPI/Code/Source/RPI.Public/View.cpp:56]: This could be an issue because if we allow view created successfully without ViewSRG get created, the view won't be work for rendering. 

Maybe we can add creating ViewSRG later in UpdateSrg() when the view is used.

Comment (VickyAtAZ) [reply Gems/Atom/RPI/Code/Source/RPI.Public/View.cpp:56]: If this condition would only happen in test, maybe we could add comment for it.

Comment (burelc-amzn) [reply Gems/Atom/RPI/Code/Source/RPI.Public/View.cpp:56]: If the `RPISystemInterface` needs to be there, then I think the test should create a mock `RPISystemInterface`. That way the production code can continue unmodified, but the test can continue to test just what it needs.

Comment (hultonha) [reply Gems/Atom/RPI/Code/Source/RPI.Public/View.cpp:56]: The problem with adding a mock is that interface is pretty huge, I mean I could do it but I'd rather factor the code to not need this whole interface to get its job done, it would be better to pass in the required dependency and have the interface be as small as possible. I _can_ add the mock but it's a load of busy work and doesn't improve the design at all (why I avoided it in this case to begin with).

File: Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp (hunk 4 above)
Comment (hultonha) [reply Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp:99]: Hmm I do most of the time but here it's not a deal breaker for me, I could rewrite this as a parameterized test but I think that add more noise to what's going on. I'd be inclined to leave this as-is for now, and if we add more tests like this in future it's easy to update and refactor

File: Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp (hunk 3 above)
Comment (hultonha) [reply Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp:92]: Good shout I can add this - so `PrintTo` would be in the header of the type and not in `MathTestHelpers`? As it's testing specific would it not make more sense there?

File: Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    {
+    public:
+        static inline constexpr AzFramework::ViewportId TestViewportId = 1234;
+        static inline AzFramework::ScreenSize ScreenDimensions = AzFramework::ScreenSize(1280, 720);
+
+        static AzFramework::ScreenPoint ScreenCenter()
+        {
+            const auto halfScreenDimensions = ScreenDimensions * 0.5f;
+            return AzFramework::ScreenPoint(halfScreenDimensions.m_width, halfScreenDimensions.m_height);
+        }
```
[the remark this answers is not in the document]
Comment (hultonha) [reply Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp:31]: I'd rather not get into adding `constexpr` all over this PR - I can add it in a few places but I don't think it's going to make a big impact to performance. We can always do this as a follow-up PR

File: Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+#include <AzTest/AzTest.h>
+#include <AzToolsFramework/UnitTest/AzToolsFrameworkTestHelpers.h>
+
+namespace UnitTest
+{
+    class ViewportInteractionImplFixture : public ::testing::Test
+    {
+    public:
+        static inline constexpr AzFramework::ViewportId TestViewportId = 1234;
+        static inline AzFramework::ScreenSize ScreenDimensions = AzFramework::ScreenSize(1280, 720);
```
[the remark this answers is not in the document]
Comment (hultonha) [reply Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp:25]: I can add this (it should be `const` anyway, that was my bad)

File: Gems/Atom/RPI/Code/Source/RPI.Public/View.cpp (hunk 5 above)
Comment (hultonha) [reply Gems/Atom/RPI/Code/Source/RPI.Public/View.cpp]: @VickyAtAZ and I discussed offline and have an update to resolve things

Comment (hultonha) [reply Gems/Atom/RPI/Code/Source/RPI.Public/View.cpp]: https://github.com/o3de/o3de/pull/5494/commits/cc0474753e13fe7f976d7b9d82873523f7e948f7

File: Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (hultonha) [reply Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp]: This is now updated 👍

File: Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp (hunk 3 above)
Comment (burelc-amzn) [reply Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp:92]: I was thinking some file that only gets included in test targets. I'm not sure if such a file exists for AzFramework currently. I wouldn't put it in the production header, since it's only going to be used in tests.

**Review by burelc-amzn — APPROVED** [review]

File: Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp (hunk 3 above)
Comment (hultonha) [reply Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp:93]: Yup that sounds good, see this commit with the relevant changes - https://github.com/o3de/o3de/pull/5494/commits/f92ee5f58cfe0363ecff10a58686cb0d3f7618a7

Comment (hultonha) [reply Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp:93]: Output now looks like this for failures

File: Gems/Atom/RPI/Code/Source/RPI.Public/View.cpp (hunk 5 above)
Comment (hultonha) [reply Gems/Atom/RPI/Code/Source/RPI.Public/View.cpp]: Getting a failure in `RenderPipelineTests, ViewFunctionsTest` with the `OnAddToRenderPipeline()` addition - it seems like it's legit for `m_shaderResourceGroup` to be null in those tests and the `AZ::Error` is causing a problem.

See here - https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-5494/4/console

Comment (hultonha) [reply Gems/Atom/RPI/Code/Source/RPI.Public/View.cpp]: The reason it fails is because the `viewSrgShaderAsset` is not ready...

Comment (hultonha) [reply Gems/Atom/RPI/Code/Source/RPI.Public/View.cpp]: As a temporary workaround as I'd really like to get this PR merged today, I've downgraded the Error to a Warning. Is this something you could then look into offline please @VickyAtAZ (https://github.com/o3de/o3de/pull/5494/commits/2c02e65437c3f44f76d6e8e3a91aaa0a52cdc3f1)

**Review by VickyAtAZ — APPROVED** [review]
